---
title: "Basic storage, access, and manipulation of phylogenetic sequencing data with *MicrobiomeExperiment*"
date: "`r Sys.Date()`"
package: MicrobiomeExperiment
output:
  BiocStyle::html_document:
    fig_height: 7
    fig_width: 10
    toc: yes
    toc_depth: 2
    number_sections: true
vignette: >
  %\VignetteIndexEntry{MicrobiomeExperiment}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: references.bib
---
`r library("knitr")`
`r opts_chunk$set(cache=FALSE, fig.width=9, message=FALSE, warning=FALSE)`

# Load *MicrobiomeExperiment*

```{r load-packages, message=FALSE, warning=FALSE}
library("MicrobiomeExperiment")
```

## Example Data

Example data consists mainly of phyloseq [[@McMurdie2013]] datasets converted to 
`SummarizedExperiment`/`TreeSummarizedExperiment` objects.

```{r, eval=FALSE}
data(GlobalPatterns)
data(esophagus)
data(enterotype)
data(soilrep) 
```

## MicrobiomeExperiment Object Summaries

Since `MicrobiomeExperiment` are `TreeSummarizedExperiment` experiment objects,
the summary should look quite familiar.

```{r}
data(GlobalPatterns)
me <- as(GlobalPatterns,"MicrobiomeExperiment")
me
```
